vault/api_redirect.go

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package vault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	url2 "net/url"
+	"strings"
+	"sync"
+
+	"github.com/armon/go-radix"
+)
+
+type apiRedirectRegistry struct {
+	lock  sync.Mutex
+	paths *radix.Tree
+}
+
+func NewAPIRedirects() *apiRedirectRegistry {
+	return &apiRedirectRegistry{
+		paths: radix.New(),
+	}
+}
+
+func (reg *apiRedirectRegistry) TryRegister(ctx context.Context, core *Core, mountUUID, src, dest string) error {
+	if strings.HasPrefix(dest, "/") {
+		return errors.New("redirect targets must be relative")
+	}
+	reg.lock.Lock()
+	defer reg.lock.Unlock()
+	_, _, found := reg.paths.LongestPrefix(src)
+	if found {
+		return fmt.Errorf("api redirect conflict for %s", src)
+	}
+	reg.paths.Insert(src, &APIRedirect{
+		c:         core,
+		mountUUID: mountUUID,
+		prefix:    dest,
+	})
+	return nil
+}
+
+func (reg *apiRedirectRegistry) Find(path string) (*APIRedirect, string) {
+	s, a, found := reg.paths.LongestPrefix(path)
+	if found {
+		remaining := strings.TrimPrefix(path, s)
+		if len(remaining) > 0 {
+			switch remaining[0] {
+			case '/':
+				remaining = remaining[1:]
+			case '?':
+			default:
+				// This isn't an exact path match
+				return nil, ""
+			}
+		}
+		return a.(*APIRedirect), remaining
+	}
+	return nil, ""
+}
+
+func (reg *apiRedirectRegistry) DeregisterMount(uuid string) {
+	reg.lock.Lock()
+	defer reg.lock.Unlock()
+	reg.paths.Walk(func(k string, v interface{}) bool {
+		r := v.(*APIRedirect)
+		if r.mountUUID == uuid {
+			reg.paths.Delete(k)
+			return true
```

Comment:
My understanding of the [WalkFn](https://pkg.go.dev/github.com/armon/go-radix#WalkFn) contract, returning true here will short circuit the iteration so if a mount has multiple values we will orphan them.

> WalkFn is used when walking the tree. Takes a key and value, returning if iteration should be terminated.

Reply:
Doh!  Good point.

Reply:
I wonder if this will break walking though.  I'll gather up the deletes first instead.